### ISSUE TITLE
On elaboration of Bookmark-search-plus-2

### ISSUE THREAD

**Issue by vitaaliy** [aaFn/Bookmark-search-plus-2#5]
This is continuation of our contact from https://www.goo.gl/fTzRWy
Hi, aaFn,
"Now, how much time (roughly) does it take to load and display your 80000 bookmark items ? And is that instantaneous with the native FF bookmark sidebar ?"

- Loading FF 58.0.1 (64 bits) in native configuration takes 6 secs. By the way, my current tabs configuration is too large - about 200 tabs (which is very bad, of course). When I switch to "Bookmark-search-plus-2" the bookmarks structure in the sidebar pane appears in 90 secs(!).

"... once displayed, you should be able to insert new bookmarks at will (or delete, move ..) = drag&drop from tabs, from links in HTML pages, or right click on a bookmark item and create/insert a new bookmark or folder or separator before it, ...just like in normal FF bookmark sidebar pane".

- When I try to drag, say, current link to some place in the bookmarks tree it looks like this operation is senseless - no visual signs. I decide that it is impossible, but after 35 secs at last appears pop-up window asking for Add-confirmation. Being confirmed - it gets into that position which I was aimed at. Deleting is done momentarily.
When I choose some folder and try to create New folder in it - it is also done with huge time lag and the new folder is created not inside, but outside the chosen one (in the same tree depth level).

"...  I ask you to open the Browser console (Ctrl+Shit+J) and send me/paste its contents, just after you try to create a new bookmark through one or more of the above methods ?"

- Okay. I took the current browser link, dragged it to some place in the pane and waited... waited... until it asked for confirmation, I agreed, the link was placed in the tree. Then I took the browser console, copied its content and here it is.
[Browser Console.txt](https://github.com/aaFn/Bookmark-search-plus-2/files/1706378/Browser.Console.txt)

**Comment by aaFn** [comment]
Mmmh I see nothing bad in the log .. I need to dig more in details what can happen here, not sure why you get no visual sign, you should at least get some blue horizontal line where you want to drag the new bookmark .. and I don't understand why the number of bookmars should impact anything on that .. will come back to you when I have time to think.

Maybe I need to create (generate) 80000 bookmarks somewhere myself for testing purposes to get a good feel for that, and also to see how to optimize the load dealay, at least visually as discussed in another issue.
On the number of tabs, no worries, I have around 300, no problem.

On creating a new folder, I didn't pay attention that indeed when we do that on a folder, it should be inside the folder and not jsut before it.
Consider it as a bug, as my intent was to mimick the native FF  Bookmarl sidebar, and I missed that particular behavior. I'll correct it.

On the

**Comment by aaFn** [comment]
Bug of inserting before folder instead of inside is corrected in 2.0.3. Let me know if not good.
I will leave that issue open because the main intent is to handle the load delay and make it more liveable for you.
And also because need to investsigate the problem of no visual sign when you insert ..

**Comment by aaFn** [comment]
Hello @vitaaliy , I made some modifications to prepare things to target your 80000 bookmarks .. :-)

Could I ask you to:
- remove the add-on if not already removed, then reinstall it ? That will clear retained state of storage, and now you will start all folders closed. Should make things more useable for you.
- then when loaded (90s later, or maybe a little faster because I tried some more marginal optimization) go to the options page, and activate the traces checkbox
- then switch the sidebar to something else, like native bookmarks, and switch again to this add-on sidebar
- 90s or so later, you will get a trace box with some performance numbers and stats ..
Can you copy that contents (ignore the error messages on cannot fetch favicon URL's if there are ..), and post them here ?

Also, on the drag and drop into the sidebar, I found some weird behavior of such actions while loading favicons .. not sure I truly cured the root cause, but on my side the phenomenon is not occurring anymore. Would you be kind enough to try again and let me know I indeed I got rid of that strange behavior ?

Thanks much in advance, aaFn.

**Comment by vitaaliy** [comment]
Hi, @aaFn,
Could I ask you to:

-     remove the add-on if not already removed, then reinstall it ? That will clear retained state of storage, and now you will start all folders closed. Should make things more useable for you.
-- Done. Yes, they start closed - still taking that same 80 secs...
-     then when loaded (90s later, or maybe a little faster because I tried some more marginal optimization) go to the options page, and activate the traces checkbox
-     then switch the sidebar to something else, like native bookmarks, and switch again to this add-on sidebar
-- Done.
-     90s or so later, you will get a trace box with some performance numbers and stats ..
-     Can you copy that contents (ignore the error messages on cannot fetch favicon URL's if there are ..), and post them here ?
-- Sure. Here it is:
PlatformOs: win
Load duration: 25223 ms
Display duration: 53151 ms
Save duration: 46 ms
fontFamily = 'Segoe UI'
fontSize   = '12px'
Stats:
------
Bookmarks:  79340
Folders:    14822
Separators: 3
--------------------
- Also, on the drag and drop into the sidebar, I found some weird behavior of such actions while loading favicons .. not sure I truly cured the root cause, but on my side the phenomenon is not occurring anymore. Would you be kind enough to try again and let me know I indeed got rid of that strange behavior ?
-- Not sure if it has anything concerning favicons... But I tried to create a subfolder somewhere in the tree structure - and now it is done OK. Than dragged the current link into the tree - the blue line mark appears correctly, then confirmation window appears - but it is done verrrrrrry slow... Irritatingly slow comparing with standard FF activity. Deleting is done quite fast. Opening or closing pane subfolders also are now lively.

Thanks much in advance, aaFn.

Thank you, aaFn! You are doing fine. The current version 2.0.4 already can be used achieving the goal formulated. Though... I think with nostalgia about that very accurate addon _"Show Parent Folder 2.1.1 By Alice0775"_. I think it was more convenient: opening _Show All Bookmarks_, very fast search and every line in result has the column with already prepared full path for the given entry - no need to follow upstairs to get it. When and how she did prepare such table - I don't know really. And I understand that now we got more unwieldy UI... but maybe you could rethink this issue once more?
Thank you again, and also all our colleagues here who actively support this project.

**Comment by aaFn** [comment]
@vitaaliy , thanks this is enlightening .. so I was seeing the same proportion on my small numbers (only 1800 bookmarks, i.e. 40 times less ..) for the load/display split durations = 1/3rd of time for load, and 2/3rd for display ... but I was not sure I could extrapolate. Now your numbers are definitely confirming :-)

Plus the stats show an average of 5 to 6 bookmarks per folder, so clearly this is well balanced in average, and so I should go for my option 2 in #2 

The fact that save is as fast on your side than on mine is a good sign that the path I am envisaging should be promising :-)

Thanks for that !


On the other topics:
- Thanks for the confirmation on drag & drop, so I really cured something but I was not sure given I could not pinpoint the root cause, and could only get the feeling / intuition this was related to fetch() calls to get favicons in the background
- Slowness = I guess I'll need to simulate your numbers to better get what is going on now, obviously the big numbers are playing a role, not sure where exactly.

Given what I envisage, I do not believe I can deliver the new way of loading in 2.0.5 .. I will prepare for it though in 2.0.5 behind the scene, and I believe the experimental async load will come just after.

Thanks again for your responsiveness.

**Comment by JohnLukeBentley** [comment]
Just to give you a baseline from another user (me) ...   

    Description: Reinstall of plugin. Load test (Side panel switch to native then back).
    Version: 2.0.5
    Datetime: 2018-02-21 20:37
    PlatformOs: win
    Load duration: 844 ms
    Display duration: 1297 ms
    Save duration: 14 ms
    fontFamily = 'Segoe UI'
    fontSize   = '12px'
    Stats:
    ---
    Bookmarks:  4191
    Folders:    1465
    Separators: 8
    ---


    Description: Reinstall of plugin. Load test (Side panel switch to native then back).
    Version: 2.0.5
    Datetime: 2018-02-21 20:39
    PlatformOs: win
    Load duration: 778 ms
    Display duration: 1093 ms
    Save duration: 3 ms
    fontFamily = 'Segoe UI'
    fontSize   = '12px'
    Stats:
    ---
    Bookmarks:  4191
    Folders:    1465
    Separators: 8
    ---

**Comment by aaFn** [comment]
@vitaaliy, @JohnLukeBentley, some bad news ..

While preparing in 2.0.6 things for optimizing the availabliity time of the display bookmark tree, which should be 2.0.7 if all goes well, by delaying the load of hidden bookmarks as described in option 2 of #2 which should have been a major breakthrough, I discovered when studying in details the API that FF does not give the separators when calling things by pieces :-(

I.e., bookmarks.getChildren() or bookmarks.get() both only return bookmarks and folders (without children which is the intended puprpose to go by pieces), but not separators ..!

See:
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/bookmarks/getSubTree
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/bookmarks/get

So to get the full details, there is only one way ... I have to stick to using bookmarks.getTree() (or bookmarks.getSubTree() but this is the same effect) which is still long in itself, waiting to get all the tree at once, and I have no way to break it in smaller pieces ..

In the above mesurements, that will mean an incompressible 25223 ms for @vitaaliy, and 778 ms for @JohnLukeBentley.

The good news is that since from experience this is around 1/3rd of the total time, that should still allow to be faster by a factor of around 3 for people with large number of bookmarks ... but that will still have a big "blank" period at start.

Also, I made a small optimization in 2.0.6 which on my side yielded a small decrease varying around 10% on display time .. don't know what the figure will be on your side, but that is always something :-)

**Comment by JohnLukeBentley** [comment]
@aaFn I have not analyzed the problem you are facing but whatever you are doing, with "Performances (not yet effective) ... Load bookmarks ..." ticked, there is an improvement on the default.

A small improvement would be to put the cursor in the Search bar when a user opens BSP2.

**Comment by aaFn** [comment]
Hello @JohnLukeBentley , I tried to get the focus to the Search bar, but didn't succeed so far ..
- autofocus on the element
- use window.focus() and set focus to that element in the page javascript
..
nothing is working so far, seems like something doesn't let me do that.

If you know of an add-on that succeeded in doing that, can you point it at me so that I look at the way it's done mimick it ?


On the side of news, I am getting better news now .. in spite of documentation, I tried .. and I can see that separators are in the answer to the call. So I am now adding the experimental "delay load" feature to 2.0.7.

**Comment by aaFn** [comment]
Good news also @vitaaliy on the delay after a drag and drop, or on insert bookmark/folder/separator, or on cut/copy and paste ...
There was indeed a bad effect induced by large numbers of rows .... not really visible below 10000 or so, but starting to show beyond that, and extremely long at 80000.

So I drastically improved the performance on that aspect, by changing the way I was proceeding for manipulating rows and remembering them for showing bookmark as search result targets.

And now this is nearly instantaneous at 80000 ... with 2.0.7 :-)

**Comment by JohnLukeBentley** [comment]
@aaFn I look forward to seeing that optimization in the next version.

On focusing ... in a pure web context I know there's this sort of thing....

https://www.w3.org/TR/html53/sec-forms.html#dom-htmlinputelement-autofocus
> Example 77
> In the following snippet, the text control would be focused when the document was loaded.
>
>      <input maxlength="256" name="q" value="" autofocus>
>      <input type="submit" value="Search">

https://www.w3schools.com/jsref/met_html_focus.asp

>       document.getElementById("myAnchor").focus(); 

... which is the sort of thing it sounds like you've tried.

**Comment by aaFn** [comment]
Correct, this is what I tried. I see that I am not alone, other people do not succeed in getting focus in their add-on .. I suppose this is a "security" feature of FF to avoid stealing focus in an unwanted way .. but this is getting very restrictive :-(
WebExtension is really being the poor parent = performance is much less than XUL, many things not possible anymore .. The world of add-ons is getting a serious step back :-(

**Comment by JohnLukeBentley** [comment]
> The world of add-ons/extensions is getting a serious step back :-(

Yes, I'm hearing that a bit. I would have thought the opportunity, taken by Mozilla, to make a clean break with legacy code would have provided add-on developers *greater* access (security restrictions aside). But I'm not familiar with the WebExtensions history. Maybe a fuller API is yet to come.

Anyway thanks for having given it a go.

**Comment by aaFn** [comment]
@vitaaliy , @JohnLukeBentley ,  News again, and sorry to blow ice and fire ... but this time **very bad** :-( 

The bookmarks.getChildren() call takes as long as the bookmarks.getSubTree() and bookmarks.getTree().
So no matter whether I get a small piece or the full piece of bookmarks, the execution duration in FF API is the same, and the cpu at 100% all long.

In short, the query looks very badly coded inside FF !!
And this is then making it inpracticable to split by pieces, since basically multiplying the full piece time by as many pieces as there are (14822 pieces or so in @vitaaliy 's case) ..

I will open a bug to FF.

Here are traces on a case where I loaded plenty of bookmarks on my FF:
```

PlatformOs: win
Setting Windows variations
Load duration: 827444 ms
Display duration: 11471 ms
fontFamily = '"Segoe UI"'
fontSize   = '12px'
Save duration: 53 ms
Stats:
------
Bookmarks:  61103
Folders:    15285
Separators: 4
--------------------


PlatformOs: win
Setting Windows variations
Root get duration: 6 ms
      Root.children: undefined
Root getChildren duration: 825553 ms
      Number of children: 4
Root getSubTree duration: 824053 ms
      Root.children: [object Object],[object Object],[object Object],[object Object]
      Number of children: 4
(Root) getTree duration: 827584 ms
      Root.children: [object Object],[object Object],[object Object],[object Object]
      Number of children: 4

```

Only "get()" is working well, but that doesn't help, because I cannot discover the tree with that, this is to get an individual piece for which I already know the id.

**And so this is the end of it for now :-(**


The only other solution would be for me to fully load everything on initial run of the add-on at install time, then store it in local storage, taking one first long blow, and then on subsequent add-on load:
- read all bookmarks from my local storage in fast mode and display that (whether all or only visible first and hidden in background)
- in background read the full bookmark tree which may have evolved on its own while BSP2 was not active (if it is active, it gets all changes and so reflects them)
- then reconcile the two together, again in background, deleting from display what is extra in BSP2 compared to the "truth", and adding what is missing.

Not ideal at all .. what are your views ?

**Comment by aaFn** [comment]
Thank you @vitaaliy for your patience and your feedback.

I should issue 2.0.7 soon wihtout the intended delay load function then (sob !), so that at least you can move things around in the tree without the 30s lags you were seeing, even if you still have to suffer from the initial long load delay ... At least that should be something more to ease your life :-)

The secret .. unfortunately they (Mozilla, FF ...) have put too much restrictions on WebExtensions and encapsulated too much .. So everything is slower or limited .. for the sake of security, but as always security is really impeding the lives of whomever it is supposed to protect, and not so often the ones we are supposed to be protected from ..

**Comment by JohnLukeBentley** [comment]
I'd join @vitaaliy in thanking you for your efforts so far. Yes it sounds like you are against the wall with Mozilla, if and until that bug is fixed.

**Comment by aaFn** [comment]
Thank you. Well, I didn't say my last word on this .. I am tempted to do the above with loading all on first run, then saving in local storage, then reconciling in background during subsequent runs. The more I think about it, the more I see value .. but I need to pause and think, because this is asking to experiment on a few things, and also because this is implying to have a different and richer object structure in the code to do it that way (which has its own benefits ..).

So pause .. but not stop ... :-)

**Comment by aaFn** [comment]
Hello @vitaaliy , 2.0.18 is out and brings what should be significant performance improvements for the operations of:
- copy & paste
- cut & paste
- move by drag & drop
- and create new bookmark

when there are lots of bookmarks.

If you pass by here, let me know if you see any improvement indeed, or nothing, or degradation ..
aaFn

**Comment by vitaaliy** [comment]
Hi @aaFn, just upgraded to 2.0.18. Initial tree loading for Bookmark-search-plus-2 2.0.18 still takes much time: 55 secs. Alas, no improvements in functionality did not appear. But I noticed very strange behavior. FF starts to labor intensively with the hard disk, and eats more and more memory during several minutes after switching 2.0.18 on. On standard bookmarks routine it takes 2.42 Gb of RAM. After switching 2.0.18 on it eats more and more - up to 7.8 Gb. I have 8 Gb RAM installed. FF tends to become unresponsive, even the mouse freeses. I returned the standard mode, tried to close FF - which it accepts very unwillingly. Then I closed its processes one after other until it died at last. This is my sad experience.
Still I appreciate your efforts with this absolutely important addon. Several times with its help I have found important folders in my huge bookmarks forest. Thank you!

**Comment by aaFn** [comment]
Ah, since how long have you not been using BSP2 ?

From the sound of what you say, it is proceeding to migration of favicons in the background, to speed up the display process in later starts.
This was introduced in 2.0.16.

What I found is that when processing images to scale them down, or when fetching them on Internet, indeed memory is growing very sharply .. However, this appears because the garbage collector waits for some time before releasing the memory.
So FF eats up all memory it can, and then at times it lowers it, or when it hits the max, it "pauses" and release memory, then continues .. but nothing really fails.

You should see that in the traces, there should be the message
`Migrating <xxx> favicons to 16x16`
appearing around 1 minute after start, see #35 and #2.

And there should be the message
`structureVersion: `
with no value yet

If this is indeed the case, then just start BSP2 before going to bed, and leave it actve alone during night. Since you have a lot of bookmarks, this is taking time.
It is migrating images at the rhythm of 5 images per second, so with your 80000 bookmarks, this should be 5 hours ..
If it crashes in the middle because of memory, not a problem, the images are saved to local storage as they get converted (hence disk working), so it will restart where it stopped last time.

That optimization is not only improving display delay (on my sytem by a factor of 2), but also the reads and writes to storage, which are part of the delay before getting the sidebar contents, and decreasing local storage a lot.

Hope this will work fine for you, else sorry about that, the 80000 bookmarks are a very uncommon volume indeed and stretch the limits.
At worst, you can go to options and simply activate the `Disable favicon fetching` option, it will discard all images and save all this .. And next time you disable it to get again the images, they will be obtained in background again, but converted on the fly which is much less intensive (and longer).

Let me know the result if you have the patience and kindness to go through that .. else I will understand anyway.

**Comment by aaFn** [comment]
@vitaaliy , @JohnLukeBentley , I measured that most of the display duration is due to favicons.
So I deferred them to after the bookmarks tree display, so that the tree can be available faster.
I also introduced some node cloning techniques which apparently are improving things a little.

Depending on conditions, I observed this is reducing the bookmarks tree display duration by a factor of 2 to 4 in my experiences. Total time of tree display duration + favicon display duration doesn't change much from before, but therefore the tree is available quite sooner.

I am putting that in 2.0.26. Together with an option to come back to previous behavior of immediate favicons display together with the tree.

At the occasion, I would be interested to see here your new measurements (those in the trace textarea) with that new version, and if things are noticeably faster (or slower ..) from your perception.

Thank you in advance if you have some time to put that here, aaFn.

**Comment by vitaaliy** [comment]
Thank you aaFn for your recent efforts! Here is my trace measurement:
==========================================
PlatformOs: win
Setting Windows variations
New jsonified save method
structureVersion: -img16-bnlist
disablefavicons_option: false
Load local store duration: 2522 ms
BSP2 version: 2.0.26
Tree load duration: 25506 ms
Tree build duration: 1210 ms
Display duration: 6247 ms
fontFamily = '"Segoe UI"'
fontSize   = '12px'
Stats:
------
Bookmarks:  81715
Folders:    15275
Separators: 3
Oddities:   0
--------------------
Save duration: 3594 ms
Favicon display duration: 17837 ms
=======================================
Besides I measured the tree loading time with my watches. It is 40 secs - surely better than before.

**Comment by aaFn** [comment]
Ok, thank you .. good to see this is having a good effect also on your side :-)

**Comment by JohnLukeBentley** [comment]
Hi @aaFn, thanks for your continued work. Two sets of measurement. Toggling `Disable favicon fetching` 

    PlatformOs: win
    Setting Windows variations
    New jsonified save method
    structureVersion: -bnlist
    disablefavicons_option: true
    Load local store duration: 84 ms
    BSP2 version: 2.0.26
    Tree load duration: 841 ms
    Tree build duration: 18 ms
    Display duration: 199 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4289
    Folders:    1496
    Separators: 8
    Oddities:   0
    --------------------
    Save duration: 116 ms
    Migrating 0 favicons to 16x16
    Nothing to migrate to 16x16 favicons
    Count of abnormal favicons: 0

    PlatformOs: win
    Setting Windows variations
    New jsonified save method
    structureVersion: -bnlist-img16
    disablefavicons_option: true
    Load local store duration: 86 ms
    BSP2 version: 2.0.26
    Tree load duration: 856 ms
    Tree build duration: 10 ms
    Display duration: 200 ms
    fontFamily = '"Segoe UI"'
    fontSize   = '12px'
    Stats:
    ------
    Bookmarks:  4289
    Folders:    1496
    Separators: 8
    Oddities:   0
    --------------------
    Save duration: 126 ms

It looks like the "Tree load duration" 841ms/856ms is unchanged from the prior "Load duration: 844 ms". But the Display duration 199ms/200 ms is considerable quicker than the prior "Display duration: 1297 ms". 

Subjectively it seems a little faster ... but I still very much feel that load duration. 

I'm looking forward to your having the tree stored in, and retrieved from, memory but I understand you might want to try general optimizations first.

**Comment by aaFn** [comment]
Thank you @JohnLukeBentley , glad to see here also that this is yielding good results.

Yes, now that display is more or less at its best, next step should be to tackle the tree load and replace API by local storage as main source, where the tree is already stored now since some time ..
I was hoping that meanwhile https://bugzilla.mozilla.org/show_bug.cgi?id=1396364 would make some progress, but no sign of it so far ..

So I will concentrate on moving to that last optimization step now. It implies transferring some code from front to the background task, which will take some time and maybe several steps for me to do safely.

I will keep you posted here and in #2.

Cheers, aaFn.

**Comment by JohnLukeBentley** [comment]
Well, thanks for plugging away at it.

Could you clarify the meaning of

    Tree load duration: 841 ms
    Tree build duration: 18 ms
    Display duration: 199 ms

Is the total time, from keyboard shortcut press to finish, the sum of all those?

**Comment by aaFn** [comment]
The total time from key press to end display should be:
```
Load local store duration: 84 ms  -> Load tree from saved file to memory
Tree load duration: 841 ms        -> Get tree from FF API
Tree build duration: 18 ms        -> This is the digestion of the tree returned by the API into the memory structure to save in file later
Display duration: 199 ms          -> This is the HTML creation and display in sidebar
```
In your case, total is now around 1150 ms, so a little more than 1s.

Your previous display duration (last step) with 2.0.5 was around 1200 ms as per one of your experiments above.
This compares now to 18 + 199 = 217 ms .. so we divided by 5 to 6.
It waa the biggest contributor so far.

Now Tree load is the biggest, and then it is the next focus area.
As you can understand, It is currently redundant with the first step, which is for now only used to retrieve the saved favicons.
So if should disappear if all goes well, as well as the digestion step just after.

We'll see if that works well .. I do not want to raise too much expectation.

**Comment by JohnLukeBentley** [comment]
Thanks @aaFn. I'd suggest adding a "Total Time" (that just sums the figures in the way you've mentioned) into the diagnostic output.

**Comment by aaFn** [comment]
Hello @JohnLukeBentley and @vitaaliy

eventually 2.0.27 is out with the next optimization step, which is to have a significant piece of the code running in background to catch any change to bookmarks even if the sidebar is not displayed, and then to read only from memory and local store when displaying.

This is yielding significant optimization when first loading FF, and even more when switching the sidebar off an on using the toolbar action button, ot the shortcut key.
When installing the add-on, this is roughly like before.

Let me know at the occasion what this is giving for you, on the numbers. (@JohnLukeBentley , I also implemented the total time).

Notes:
1. It appears that with 61.0.1, the FF API takes longer to return with the bookmark tree .. by around 15 to 20% from my measurements. This may be due to https://bugzilla.mozilla.org/show_bug.cgi?id=1472127
2. I hope that for big numbers like yours, there is no bad side effect of having the process in background active all the time on bookmarks, in order to catch all changes and to keep the tree up to date in memory for fast availablity.
Let me know ..

Thanks, aaFn.

**Comment by vitaaliy** [comment]
Hi aaFn,
Bad news now. 2.0.27 does not work at all. It hangs on _Load FF APL..._ endlessly. For an hour. I tried to Activate traces but no traces appeared. Thinking now about returning the previous version - it worked and gave me considerable help.

**Comment by vitaaliy** [comment]
Hi aaFn,
Bad news now. 2.0.27 does not work at all. It hangs on _Load FF APL..._ endlessly. For an hour. I tried to Activate traces but no traces appeared. Thinking now about returning the previous version - it worked and gave me considerable help.

**Comment by vitaaliy** [comment]
Hmmm... My next try was to make cleaner experiment. I deleted the current version, reloaded FF and installed it anew. And it works now! And I managed to get traces:
PlatformOs: win
Setting Windows variations
Load local store duration: 71 ms
Background load local store duration: 45 ms
Background tree load duration: 30214 ms
Background tree build duration: 259 ms
Background save duration: 9720 ms
Get tree duration: 0 ms
structureVersion: -img16-bnlist
disableFavicons_option: false
Display duration: 8266 ms
Total duration: 8337 ms
fontFamily = '"Segoe UI"'
fontSize   = '12px'
Stats:
------
Bookmarks:  164290
Folders:    30693
Separators: 6
Oddities:   0
--------------------
BSP2 version: 2.0.27
Favicon display duration: 23857 ms
-------------------------------------------------------------------
The strange thing is that it says - Bookmarks:  164290 … As we kow, recently I had about 84,000. Probably it doubles the number...
I tried to switch to native Bms and back to BSP - and found that it really takes considerably less time than with previous version.
Thank you!

**Comment by aaFn** [comment]
Hello @vitaaliy , on the hang on "_Load FF APL..._" my bad .. I interpreted a little wrongly how to use browser.runtime.onInstalled.addListener() https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/runtime/onInstalled

No event is sent on it when FF regularly restarts, but the code was waiting on it to systematically arrive in order to decide if we need to reload bookmarks from the FF API to resync or not .. and so it was hanging on normal FF restarts. I got the problem also.

This was working perfectly during test because the addon was tested as a temporary addon, and this event alway gets fired in such a mode ...

This should be cured now with 2.0.28 just out .. Let me know if you still have troubles ... don't use 2.0.27 then :-)   Sorry.

On double bookmarks count, that is strange.
Would you mind refreshing to 2.0.28, and then:
- open the browser console,
- go to the options panel,
- and press the `Reload from FF API now` button
- tell me if you still see a double count and the contents of the traces + browser console ?

Thanks, aaFn.

**Comment by vitaaliy** [comment]
Hi aaFn,
BSP2 upgraded to 2.0.28 automatically. I opened the browser console: Ctrl+Shift+J but could not find "options" panel there. Then I activated traces and switched from natives to BSP2. After 10 secs I got your tree :). Here are the traces:
PlatformOs: win
Setting Windows variations
Load local store duration: 106 ms
Background load local store duration: 4283 ms
Background bypassed FF API for tree load
Background tree load duration: 457 ms
Background tree build duration: 60 ms
Background save duration: 2201 ms
Get tree duration: 1 ms
structureVersion: -img16-bnlist
disableFavicons_option: false
Display duration: 8138 ms
Total duration: 8245 ms
fontFamily = '"Segoe UI"'
fontSize   = '12px'
Stats:
------
**_Bookmarks:  82150_**
Folders:    15349
Separators: 3
Oddities:   0
--------------------
BSP2 version: 2.0.28
Favicon display duration: 18863 ms

So, results obtained are much better now. And the number of Bookmarks looks realistically. Thank you again. Good job indeed!

**Comment by aaFn** [comment]
Ok, good thanks, looks like we are back on track then :-)
So around 10s now, compared to around 50s in the past .. indeed that is a big effect. Happy to see that.

I believe we should consider closing that issue now, we're at the bottom of what could reasonably be done. Let's wait a little more to see if all remains good, and how are things gonig for @JohnLukeBentley .. and then if all ok I will close.

Thanks aaFn.

**Comment by aaFn** [comment]
@JohnLukeBentley documented the results for him in #2, they are also positive.
All look s good, and stable now, so I am closing that issue.
As said elsewhere, if we need one day to introduce a further optimization on display by displaying first only the visible parts, I will open a new issue as it will be a complex one requiring its own thread.

Thank you for all the feedback, aaFn.
